proto-lens/src/Data/ProtoLens/Encoding/Wire.hs

```diff
@@ -1,29 +1,112 @@
 -- | Module defining the individual base wire types (e.g. VarInt, Fixed64).
 --
--- TODO: separate out this definition from the Encoding implementation.
--- Also consider moving away from a GADT to a simpler sum type.
+-- They are used to represent the @unknownFields@ within the proto message.
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 module Data.ProtoLens.Encoding.Wire
     ( Tag(..)
     , TaggedValue(..)
-    , WireType(..)
     , WireValue(..)
-    , SomeWireType(..)
-    , parseTaggedValue
-    , buildTaggedValue
+    , FieldSet
+    , splitTypeAndTag
+    , joinTypeAndTag
+    , parseFieldSet
+    , buildFieldSet
+    , parseTaggedValueFromWire
     ) where
 
-import Data.Word (Word64)
+import Control.DeepSeq (NFData(..))
+import Data.Bits ((.&.), (.|.), shiftL, shiftR)
+import qualified Data.ByteString as B
+import Data.Word (Word8, Word32, Word64)
 
 import Data.ProtoLens.Encoding.Bytes
-import Data.ProtoLens.Encoding.Reflected.Wire
 
--- | Parse an unknown WireValue, given that the type+number has
--- already been parsed.
-parseTaggedValue :: Word64 -> Parser TaggedValue
-parseTaggedValue w = do
-    (SomeWireType wt, tag) <- runEither $ splitTypeAndTag w
-    value <- getWireValue wt
-    return $ TaggedValue tag (WireValue wt value)
+-- | A tag that identifies a particular field of the message when converting
+-- to/from the wire format.
+newtype Tag = Tag { unTag :: Int}
+    deriving (Show, Eq, Ord, Num, NFData)
+
+-- | The encoding of some unknown field on the wire.
+data WireValue
+    = VarInt !Word64
+    | Fixed64 !Word64
+    | Lengthy !B.ByteString
+    | StartGroup
+    | EndGroup
+    | Fixed32 !Word32
+    deriving (Eq, Ord)
+
+-- | A pair of an encoded field and a value.
+data TaggedValue = TaggedValue !Tag !WireValue
+    deriving (Eq, Ord)
+
+type FieldSet = [TaggedValue]
+
+-- TaggedValue, WireValue and Tag are strict, so their NFData instances are
+-- trivial:
+instance NFData TaggedValue where
+    rnf = (`seq` ())
+
+instance NFData WireValue where
+    rnf = (`seq` ())
 
 buildTaggedValue :: TaggedValue -> Builder
-buildTaggedValue = putTaggedValue
+buildTaggedValue (TaggedValue tag wv) =
+    putVarInt (joinTypeAndTag tag (wireValueToInt wv))
+    <> buildWireValue wv
+
+buildWireValue :: WireValue -> Builder
+buildWireValue (VarInt w) = putVarInt w
+buildWireValue (Fixed64 w) = putFixed64 w
+buildWireValue (Fixed32 w) = putFixed32 w
+buildWireValue (Lengthy b) =
+    putVarInt (fromIntegral $ B.length b)
+    <> putBytes b
+buildWireValue StartGroup = mempty
+buildWireValue EndGroup = mempty
+
+wireValueToInt :: WireValue -> Word8
+wireValueToInt VarInt{} = 0
```

Comment:
If this encoding corresponds to anything external, link to the spec?

Reply:
Done.